test/e2e/utils_test.go

```diff
@@ -327,3 +330,26 @@ func verifyCertificateRenewed(ctx context.Context, secretName, namespace string,
 		return true, nil
 	})
 }
+
+// create randomized string
+func randomStr(size int) string {
+	char := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+	rand.NewSource(time.Now().UnixNano())
+	var s bytes.Buffer
+	for i := 0; i < size; i++ {
+		s.WriteByte(char[rand.Int63()%int64(len(char))])
+	}
+	return s.String()
+}
+
+// replace string in file
+func replaceStr(oldStr string, newStr string, fileName string) error {
+	f, err := ioutil.ReadFile(fileName)
```

Comment:
We should read from `testassets.ReadFile` here instead of `ioutil.ReadFile` in order to ensure that this file can be read at runtime directly from `testassets` (embedded FS) instead of having to require the file be present at test/run time.

Alternatively, it should also be good to take the file reader function as an argument to this function something like `func replaceStr(oldStr string, newStr string, assetFunc func(name string) ([]byte, error), fileName string) { ... }` very similar to what is done in `loader.CreateFromFile` function.

Reply:
@swghosh There is write file step in func: replaceStr()
err = ioutil.WriteFile(fileName, []byte(f1), 0644)
but it seems go embed.FS only support ReadFile, do you have any experience?

Reply:
As [embed.FS](https://pkg.go.dev/embed) is a read-only file system it doesn't support write mechanisms.
For your specific case, I see that write file is explicitly not required.

Reply:
My suggestion is something similar to:

`$ git apply modified.patch`

```diff
diff --git a/test/e2e/certificates_test.go b/test/e2e/certificates_test.go
index 440bb2c7..43df958b 100644
--- a/test/e2e/certificates_test.go
+++ b/test/e2e/certificates_test.go
@@ -6,6 +6,7 @@ package e2e
 import (
 	"context"
 	"path/filepath"
+	"strings"
 	"time"
 
 	corev1 "k8s.io/api/core/v1"
@@ -281,7 +282,7 @@ var _ = Describe("ACME Certificate", Ordered, func() {
 
 			By("Check cloud credential in cluster")
 			_, err = loader.KubeClient.CoreV1().Secrets("kube-system").Get(ctx, "gcp-credentials", metav1.GetOptions{})
-			if err != nil {			
+			if err != nil {
 				Skip("Skipping for the cluster without credential in cluster")
 			}
 
@@ -316,17 +317,20 @@ var _ = Describe("ACME Certificate", Ordered, func() {
 			defer loader.DeleteTestingNS(ns.Name)
 
 			By("Creating new certificate")
-			certFile := filepath.Join("testdata", "acme", "certificate_gcp.yaml")
-			randomString := randomStr(7)
-			err = replaceStr("RANDOM_STR", randomString, certFile)
-			Expect(err).NotTo(HaveOccurred())
-			defer func() {
-				err = replaceStr(randomString, "RANDOM_STR", certFile)
-	                        Expect(err).NotTo(HaveOccurred())
-			}()
+			loadFileAndReplaceStr := func(fileName string) ([]byte, error) {
+				bytes, err := testassets.ReadFile(fileName)
+				if err != nil {
+					return nil, err
+				}
+
+				fileContentsStr := string(bytes)
+				randomString := randomStr(7)
+				fileContentsStr = strings.ReplaceAll(fileContentsStr, "RANDOM_STR", randomString)
+				return []byte(fileContentsStr), nil
+			}
 			// The name is defined by the testdata YAML file certificate_gcp.yaml
 			certName := "cert-with-acme-dns01-clouddns-ambient"
-			loader.CreateFromFile(testassets.ReadFile, filepath.Join("testdata", "acme", "certificate_gcp.yaml"), ns.Name)
+			loader.CreateFromFile(loadFileAndReplaceStr, filepath.Join("testdata", "acme", "certificate_gcp.yaml"), ns.Name)
 			By("Waiting for certificate to get ready")
 			err = waitForCertificateReadiness(ctx, certName, ns.Name)
 			Expect(err).NotTo(HaveOccurred())
diff --git a/test/e2e/utils_test.go b/test/e2e/utils_test.go
index cd696847..c0832683 100644
--- a/test/e2e/utils_test.go
+++ b/test/e2e/utils_test.go
@@ -8,7 +8,6 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
-	"io/ioutil"
 	"math/rand"
 	"strings"
 	"sync"
@@ -341,15 +340,3 @@ func randomStr(size int) string {
 	}
 	return s.String()
 }
-
-// replace string in file
-func replaceStr(oldStr string, newStr string, fileName string) error {
-	f, err := ioutil.ReadFile(fileName)
-	if err != nil {
-		return err
-	}	
-	f1 := strings.ReplaceAll(string(f), oldStr, newStr)
-	err = ioutil.WriteFile(fileName, []byte(f1), 0644)
-	return err
-}
-

```